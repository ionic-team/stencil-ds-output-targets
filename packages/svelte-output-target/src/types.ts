export interface OutputTargetSvelte {
  componentCorePackage?: string;
  legacy?: boolean;
  proxiesFile: string;
  excludeComponents?: string[];
  accessors?: boolean;
  componentBindings?: ComponentBindingConfig[];
  loaderDir?: string;
  includePolyfills?: boolean;
  includeDefineCustomElements?: boolean;
}

export interface ComponentBindingConfig {
  elements: string | string[];
  event: string;
  targetProp: string;
}

export interface PackageJSON {
  types: string;
}
