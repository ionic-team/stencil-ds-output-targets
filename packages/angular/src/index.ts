export { angularOutputTarget } from './plugin';
export type { OutputTargetAngular, ValueAccessorConfig } from './types';
