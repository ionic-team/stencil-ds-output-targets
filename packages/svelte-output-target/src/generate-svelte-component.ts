import type { ComponentCompilerMeta } from '@stencil/core/internal';
import { ComponentBindingConfig } from './types';

export const createComponentDefinition = (
  cmpMeta: Pick<ComponentCompilerMeta, 'properties' | 'tagName' | 'methods' | 'events'>,
  bindingsConfig?: ComponentBindingConfig[],
) => {
  const {
    tagName, properties, methods, events,
  } = cmpMeta;

  const bindings = bindingsConfig
    ?.filter((c) => (
      Array.isArray(c.elements) ? c.elements.includes(tagName) : (c.elements === tagName)
    ))
    .filter((c1, index, self) => index === self.findIndex((c2) => (
      (c1.event === c2.event) && (c1.targetProp === c2.targetProp)
    )))
    .map((c) => `if (e.type === '${c.event}') { ${c.targetProp} = e.detail; }`)
    .join('\n  ');

  return `
<script>
import { createEventDispatcher, onMount } from 'svelte';

let __ref;
let __mounted = false;

const dispatch = createEventDispatcher();

${properties.map((prop) => `export let ${prop.name}${!prop.required ? ' = undefined' : ''};`).join('\n')}

${
  methods
    .map((method) => `export const ${method.name} = (...args) => __ref.${method.name}(...args);`)
    .join('\n')
}

export const getWebComponent = () => __ref;

onMount(() => { __mounted = true; });

const setProp = (prop, value) => { if (__ref) __ref[prop] = value; };

${
  properties
    .filter((prop) => !prop.attribute)
    .map((prop) => `$: if (__mounted) setProp('${prop.name}', ${prop.name});`)
    .join('\n')
}

const onEvent = (e) => {
  e.stopPropagation();
  dispatch(e.type, e.detail);${bindings ? `\n  ${bindings}` : ''}
};
</script>

<${tagName} 
  ${properties
    .filter((prop) => !!prop.attribute)
    .map((prop) => `${prop.attribute}={${prop.name}}`).join('\n  ')}
  ${events
    .map((event) => `on:${event.name}={onEvent}`).join('\n  ')}
  bind:this={__ref}
>
  <slot></slot>
</${tagName}>
  `;
};
