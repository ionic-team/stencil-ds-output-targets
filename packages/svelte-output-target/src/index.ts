export { svelteOutputTarget } from './plugin';
export type { OutputTargetSvelte, ComponentBindingConfig } from './types';
