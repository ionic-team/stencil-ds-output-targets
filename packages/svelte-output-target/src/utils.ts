/* eslint-disable no-param-reassign */

import path from 'path';
import { promisify } from 'util';
import fs from 'fs';
import type { PackageJSON } from './types';

const readFile = promisify(fs.readFile);

const EXTENDED_PATH_REGEX = /^\\\\\?\\/;
const SLASH_REGEX = /\\/g;
// eslint-disable-next-line no-control-regex
const NON_ASCII_REGEX = /[^\x00-\x80]+/;

export const toLowerCase = (str: string) => str.toLowerCase();

export const dashToCamelCase = (str: string) => str
  .replace(/-([a-z])/g, (_, up) => up.toUpperCase());

export const dashToPascalCase = (str: string) => toLowerCase(str)
  .split('-')
  .map((segment) => segment.charAt(0).toUpperCase() + segment.slice(1))
  .join('');

export const sortBy = <T>(array: T[], prop: (item: T) => string) => array.slice().sort((a, b) => {
  const nameA = prop(a);
  const nameB = prop(b);
  if (nameA < nameB) return -1;
  if (nameA > nameB) return 1;
  return 0;
});

export function normalizePath(str: string) {
  if (typeof str !== 'string') {
    throw new Error('invalid path to normalize');
  }

  str = str.trim();

  if (EXTENDED_PATH_REGEX.test(str) || NON_ASCII_REGEX.test(str)) {
    return str;
  }

  str = str.replace(SLASH_REGEX, '/');

  // always remove the trailing /
  // this makes our file cache look ups consistent
  if (str.charAt(str.length - 1) === '/') {
    const colonIndex = str.indexOf(':');
    if (colonIndex > -1) {
      if (colonIndex < str.length - 2) {
        str = str.substring(0, str.length - 1);
      }
    } else if (str.length > 1) {
      str = str.substring(0, str.length - 1);
    }
  }

  return str;
}

export const relativeImport = (pathFrom: string, pathTo: string, ext?: string) => {
  let relativePath = path.relative(path.dirname(pathFrom), path.dirname(pathTo));
  if (relativePath === '') {
    relativePath = '.';
  } else if (relativePath[0] !== '.') {
    relativePath = `./${relativePath}`;
  }
  return normalizePath(`${relativePath}/${path.basename(pathTo, ext)}`);
};

export const readPackageJson = async (rootDir: string) => {
  const pkgJsonPath = path.join(rootDir, 'package.json');

  let pkgJson: string;
  try {
    pkgJson = await readFile(pkgJsonPath, 'utf8');
  } catch (e) {
    throw new Error(`Missing "package.json" file for distribution: ${pkgJsonPath}`);
  }

  let pkgData: PackageJSON;
  try {
    pkgData = JSON.parse(pkgJson);
  } catch (e) {
    throw new Error(`Error parsing package.json: ${pkgJsonPath}, ${e}`);
  }

  return pkgData;
};
