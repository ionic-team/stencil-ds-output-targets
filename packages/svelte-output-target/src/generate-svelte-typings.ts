import { ComponentCompilerMeta } from '@stencil/core/internal';
import { dashToPascalCase } from './utils';

export const generateTypings = (meta: ComponentCompilerMeta) => {
  const name = dashToPascalCase(meta.tagName);
  const jsxEventName = (eventName: string) => `on${eventName.charAt(0).toUpperCase() + eventName.slice(1)}`;

  const types = `
interface ${name}Props {
  ${
  meta.properties
    .map((prop) => `\n  /** ${prop.docs.text} */\n  ${prop.name}?: Components.${name}["${prop.name}"]`)
    .join('\n  ')}
}

interface ${name}Events {
  ${
  meta.events
    .map((event) => `\n  /** ${event.docs.text} */\n  ${event.name}: Parameters<JSX.${name}["${jsxEventName(event.name)}"]>[0]`)
    .join('\n  ')}
}

interface ${name}Slots {
  default: any
}
  `;

  return types;
};

export const generate$$TypeDefs = (meta: ComponentCompilerMeta, source: string) => {
  const name = dashToPascalCase(meta.tagName);

  const inject = [
    'extends SvelteComponent {',
    // For some reason n-1 $ signs appear in output...?
    `$$$prop_def: ${name}Props;`,
    `$$$events_def: ${name}Events;`,
    `$$$slot_def: ${name}Slots;\n`,
    `$on<K extends keyof ${name}Events>(type: K, callback: (e: ${name}Events[K]) => any): () => void {\n\t  return super.$on(type, callback);\n\t}\n`,
    `$set($$$props: Partial<${name}Props>): void {\n\t  super.$set($$$props);\n\t}\n`,
  ].join('\n  ');

  return source.replace('extends SvelteComponent {', inject);
};

export const replaceMethodDefs = (meta: ComponentCompilerMeta, source: string) => {
  const name = dashToPascalCase(meta.tagName);

  let newSource = source;

  newSource = source.replace('get getWebComponent() {', `get getWebComponent(): HTML${name}Element | undefined {`);

  meta
    .methods
    .forEach((method) => {
      newSource = newSource
        .replace(`get ${method.name}() {`, `\n  /** ${method.docs.text} */\n get ${method.name}(): Components.${name}["${method.name}"] {`);
    });

  return newSource;
};
