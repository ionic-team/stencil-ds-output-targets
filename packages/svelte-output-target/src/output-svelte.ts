import path from 'path';
import type {
  CompilerCtx, ComponentCompilerMeta, Config, OutputTargetDist,
} from '@stencil/core/internal';
import { OutputTargetSvelte } from './types';
import { sortBy, normalizePath, dashToPascalCase } from './utils';
import { createComponentDefinition } from './generate-svelte-component';
import { generate$$TypeDefs, generateTypings, replaceMethodDefs } from './generate-svelte-typings';

const svelte = require('svelte/compiler');

const REGISTER_CUSTOM_ELEMENTS = 'defineCustomElements';
const APPLY_POLYFILLS = 'applyPolyfills';
const DEFAULT_LOADER_DIR = '/dist/loader';

const ignoreChecks = () => [
  '/* eslint-disable */',
  '/* tslint:disable */',
  '// @ts-nocheck',
].join('\n');

const getFilteredComponents = (
  excludeComponents: string[] = [],
  cmps: ComponentCompilerMeta[],
) => sortBy<ComponentCompilerMeta>(
  cmps,
  (cmp: ComponentCompilerMeta) => cmp.tagName,
).filter(
  (c: ComponentCompilerMeta) => !excludeComponents.includes(c.tagName) && !c.internal,
);

export const getPathToCorePackageLoader = (config: Config, outputTarget: OutputTargetSvelte) => {
  const basePkg = outputTarget.componentCorePackage || '';
  const distOutputTarget = config.outputTargets?.find((o) => o.type === 'dist') as OutputTargetDist;

  const distAbsEsmLoaderPath = (distOutputTarget?.esmLoaderPath
    && path.isAbsolute(distOutputTarget.esmLoaderPath))
    ? distOutputTarget.esmLoaderPath
    : null;

  const distRelEsmLoaderPath = config.rootDir && distAbsEsmLoaderPath
    ? path.relative(config.rootDir, distAbsEsmLoaderPath)
    : null;

  const loaderDir = outputTarget.loaderDir || distRelEsmLoaderPath || DEFAULT_LOADER_DIR;
  return normalizePath(path.join(basePkg, loaderDir));
};

export function generateProxies(
  config: Config,
  components: ComponentCompilerMeta[],
  outputTarget: OutputTargetSvelte,
) {
  const pathToCorePackageLoader = getPathToCorePackageLoader(config, outputTarget);

  let sourceImports = `import '${pathToCorePackageLoader}';`;
  let registerCustomElements = '';

  if (outputTarget.includePolyfills && outputTarget.includeDefineCustomElements) {
    sourceImports = `import { ${APPLY_POLYFILLS}, ${REGISTER_CUSTOM_ELEMENTS} } from '${pathToCorePackageLoader}';\n`;
    registerCustomElements = `${APPLY_POLYFILLS}().then(() => ${REGISTER_CUSTOM_ELEMENTS}());`;
  } else if (!outputTarget.includePolyfills && outputTarget.includeDefineCustomElements) {
    sourceImports = `import { ${REGISTER_CUSTOM_ELEMENTS} } from '${pathToCorePackageLoader}';\n`;
    registerCustomElements = `${REGISTER_CUSTOM_ELEMENTS}();`;
  }

  const fileName = (c: ComponentCompilerMeta) => dashToPascalCase(c.tagName);

  const buildImports = (
    dir: string,
    ext = '',
  ) => components.map((c) => `import ${fileName(c)} from '${dir}/${fileName(c)}${ext}';`).join('\n');

  const buildExports = () => components.map((c) => `export { ${fileName(c)} };`).join('\n');

  const entry = [
    ignoreChecks(),
    sourceImports,
    registerCustomElements,
    buildImports('./components'),
    buildExports(),
  ].join('\n');

  const uncompiledFiles = components.map((c) => ({
    name: fileName(c),
    meta: c,
    content: createComponentDefinition(c, outputTarget.componentBindings),
  }));

  const uncompiledEntry = [
    ignoreChecks(),
    buildImports('.', '.svelte'),
    buildExports(),
  ].join('\n');

  const compiledFiles = uncompiledFiles.map((file) => ({
    name: file.name,
    meta: file.meta,
    content: svelte.compile(file.content, {
      name: file.name,
      legacy: outputTarget.legacy,
      accessors: outputTarget.accessors,
      css: false,
      preserveComments: true,
      outputFilename: file.name,
    }).js.code,
  }));

  return {
    entry,
    uncompiledEntry,
    uncompiledFiles,
    compiledFiles,
  };
}

export const svelteProxyOutput = async (
  config: Config,
  compilerCtx: CompilerCtx,
  outputTarget: OutputTargetSvelte,
  components: ComponentCompilerMeta[],
) => {
  const filteredComponents = getFilteredComponents(outputTarget.excludeComponents, components);

  const output = generateProxies(config, filteredComponents, outputTarget);
  await compilerCtx.fs.writeFile(outputTarget.proxiesFile, output.entry);
  const outputDir = path.dirname(outputTarget.proxiesFile);
  const uncompiledDir = path.resolve(outputDir, 'svelte');
  const compiledDir = path.resolve(outputDir, 'components');

  await compilerCtx.fs.writeFile(path.resolve(uncompiledDir, 'index.js'), output.uncompiledEntry);

  await Promise.all(output.uncompiledFiles.map((file) => {
    const filePath = path.resolve(uncompiledDir, `${file.name}.svelte`);
    return compilerCtx.fs.writeFile(filePath, file.content);
  }));

  await Promise.all(output.compiledFiles.map((file) => {
    const filePath = path.resolve(compiledDir, `${file.name}.ts`);
    const { content, meta } = file;

    return compilerCtx.fs.writeFile(
      filePath,
      [
        ignoreChecks(),
        `import { Components, JSX } from '${outputTarget.componentCorePackage}';\n`,
        generateTypings(meta),
        replaceMethodDefs(meta, generate$$TypeDefs(meta, content)),
      ].join('\n'),
    );
  }));
};
