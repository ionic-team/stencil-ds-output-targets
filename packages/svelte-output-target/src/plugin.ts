import type { Config, OutputTargetCustom } from '@stencil/core/internal';
import path from 'path';
import { normalizePath } from './utils';
import type { OutputTargetSvelte } from './types';
import { svelteProxyOutput } from './output-svelte';

export const normalizeOutputTarget = (config: Config, outputTarget: any) => {
  const results: OutputTargetSvelte = {
    ...outputTarget,
    excludeComponents: outputTarget.excludeComponents || [],
    componentModels: outputTarget.componentModels || [],
    includePolyfills: outputTarget.includePolyfills ?? true,
    includeDefineCustomElements: outputTarget.includeDefineCustomElements ?? true,
  };

  if (config.rootDir == null) {
    throw new Error('rootDir is not set and it should be set by stencil itself');
  }

  if (outputTarget.proxiesFile == null) {
    throw new Error('proxiesFile is required');
  }

  if (outputTarget.directivesProxyFile && !path.isAbsolute(outputTarget.directivesProxyFile)) {
    results.proxiesFile = normalizePath(path.join(config.rootDir, outputTarget.proxiesFile));
  }

  return results;
};

export const svelteOutputTarget = (outputTarget: OutputTargetSvelte): OutputTargetCustom => ({
  type: 'custom',
  name: 'svelte-library',
  validate(config) {
    return normalizeOutputTarget(config, outputTarget);
  },
  async generator(config, compilerCtx, buildCtx) {
    const timespan = buildCtx.createTimeSpan('generate svelte started', true);

    await svelteProxyOutput(config, compilerCtx, outputTarget, buildCtx.components);

    timespan.finish('generate svelte finished');
  },
});
