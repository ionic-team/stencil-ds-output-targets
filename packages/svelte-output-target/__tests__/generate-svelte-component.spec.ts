import { createComponentDefinition } from '../src/generate-svelte-component';

describe('createComponentDefinition', () => {
  it('should create a Svelte component', () => {
    const output = createComponentDefinition({
      tagName: 'my-component',
      properties: [{ 
        name: 'propOne', 
        attribute: 'prop-one' 
      }, { 
        name: 'propTwo',
        attribute: 'prop-two' 
      }, { 
        name: 'propThree',
      }],
      methods: [{ name: 'methodOne' }, { name: 'methodTwo' }, { name: 'methodThree' }],
      events: [{ name: 'eventOne' }, { name: 'eventTwo' }, { name: 'eventThree' }],
    } as any);

    expect(output).toMatchSnapshot();
  });
});
