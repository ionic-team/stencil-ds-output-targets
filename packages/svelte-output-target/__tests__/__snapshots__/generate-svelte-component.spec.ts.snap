// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createComponentDefinition should create a Svelte component 1`] = `
"
<script>
import { createEventDispatcher, onMount } from 'svelte';

let __ref;
let __mounted = false;

const dispatch = createEventDispatcher();

export let propOne = undefined;
export let propTwo = undefined;
export let propThree = undefined;

export const methodOne = (...args) => __ref.methodOne(...args);
export const methodTwo = (...args) => __ref.methodTwo(...args);
export const methodThree = (...args) => __ref.methodThree(...args);

export const getWebComponent = () => __ref;

onMount(() => { __mounted = true; });

const setProp = (prop, value) => { if (__ref) __ref[prop] = value; };

$: if (__mounted) setProp('propThree', propThree);

const onEvent = (e) => {
  e.stopPropagation();
  dispatch(e.type, e.detail);
};
</script>

<my-component 
  prop-one={propOne}
  prop-two={propTwo}
  on:eventOne={onEvent}
  on:eventTwo={onEvent}
  on:eventThree={onEvent}
  bind:this={__ref}
>
  <slot></slot>
</my-component>
  "
`;
